### PR TITLE
Implement REST API for applications discoverable for user.

### PR DESCRIPTION
**Pull request by malithie**
## Purpose
Resolves wso2/product-is#6801

## Goals
Adds the REST API implementation for applications discoverable for end users.

## Approach
- Adds API definition for Open API
- Implements API definition

## User stories
Refer https://github.com/wso2/product-is/issues/6794

## Documentation
https://github.com/wso2/product-is/issues/6807

## Automation tests
 - Integration tests
   https://github.com/wso2/product-is/issues/6808

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
https://github.com/wso2/product-is/issues/6809

## Related PRs
https://github.com/wso2/carbon-identity-framework/pull/2520

## Migrations (if applicable)
https://github.com/wso2/product-is/issues/6810

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.framework</groupId>
+            <artifactId>org.wso2.carbon.identity.application.mgt</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+
+
+</project>
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml]: new line

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            <artifactId>spring-web</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.framework</groupId>
+            <artifactId>org.wso2.carbon.identity.application.mgt</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml]: unnecessary empty line

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# OpenAPI Generator Ignore
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore]: I think this file is not needed.

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/pom.xml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.framework</groupId>
+            <artifactId>org.wso2.carbon.identity.application.mgt</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+    
+</project>
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/pom.xml]: new line

File: components/org.wso2.carbon.identity.api.user.application/pom.xml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    <artifactId>org.wso2.carbon.identity.api.user.application</artifactId>
+    <packaging>pom</packaging>
+    <modules>
+        <module>org.wso2.carbon.identity.api.user.application.common</module>
+        <module>org.wso2.carbon.identity.rest.api.user.application.v1</module>
+    </modules>
+
+
+</project>
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.api.user.application/pom.xml]: new line

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore (hunk 3 above)
Comment (malithie) [reply components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore]: This file is something similar to gitignore.
Specifically committed as we should avoid regenerating the service impl at /impl path

File: components/org.wso2.carbon.identity.api.user.application/pom.xml (hunk 5 above)
Comment (malithie) [reply components/org.wso2.carbon.identity.api.user.application/pom.xml]: fixed

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/pom.xml (hunk 4 above)
Comment (malithie) [reply components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/pom.xml]: fixed

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml (hunk 2 above)
Comment (malithie) [reply components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml]: fixed

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml (hunk 1 above)
Comment (malithie) [reply components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.api.user.application.common/pom.xml]: fixed

File: components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore (hunk 3 above)
Comment (madurangasiriwardena) [reply components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore]: Yes. But the added file only has commented items :)

Comment (malithie) [reply components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore]: Not the last line. I will update the file such that only necessary content is there

Comment (malithie) [reply components/org.wso2.carbon.identity.api.user.application/org.wso2.carbon.identity.rest.api.user.application.v1/.openapi-generator-ignore]: Fixed

**Pull request merged**